src/librustc/traits/engine.rs

```diff
@@ -15,7 +15,7 @@ use hir::def_id::DefId;
 use super::{FulfillmentContext, FulfillmentError};
 use super::{ObligationCause, PendingPredicateObligation, PredicateObligation};
 
-pub trait TraitEngine<'tcx> {
+pub trait TraitEngine<'tcx>: 'tcx {
```

Comment:
@nikomatsakis  it is the first time that I encounter lifetime bound like this, I usually see `&'a T` or `<'a>`/`<T: 'a>`, and it seems this syntax isn't covered in [[RFC-0599]-Default Object Bound](https://github.com/nox/rust-rfcs/blob/master/text/0599-default-object-bound.md).
I tried this on a struct:
```
struct Test<'a>: 'a {
    s: &'a str,
}
```
it doesn't work, so is this syntax specific to `Trait`?

Reply:
@csmoe 

> so is this syntax specific to `Trait`?

yes and no. In general, when you write `trait Foo: Bar`, that is short for `trait Foo where Self: Bar`. This is no exception; `trait TraitEngine<'tcx>: 'tcx` is short for

```rust
trait TraitEngine<'tcx>
where Self: 'tcx
{ ... }
```

The `where T: 'a` notation is generally applicable (e.g., `struct Foo<T> where T: 'a`). It means "all references in the type `T` outlive `'a`".